### PR TITLE
Temporary intervals

### PR DESCRIPTION
**Pull request by cosunae**
Description
=============

Captures the intervals where temporaries are accessed in order to do a proper allocation size of storages in the code generation

Changes: 
* Fix a memory issue with shared_ptr being extracted (OptimizerContext) as raw pointers. When the shared_ptr was out of scope the whole StencilInstantiation was deallocated, while raw pointers still used. 
* Compute the intervals per field being accessed in the Stage::update()
* Unittest the interval computation

### CODE REVIEW

**Comment by cosunae** [comment]
launch jenkins

File: src/dawn/Optimizer/StencilInstantiation.cpp [hunk 1]
```diff
@@ -995,10 +995,10 @@ class StencilDescStatementMapper : public ASTVisitor {
 //===------------------------------------------------------------------------------------------===//
 
 StencilInstantiation::StencilInstantiation(OptimizerContext* context,
-                                           const sir::Stencil* SIRStencil, const SIR* SIR)
+                                           const std::shared_ptr<sir::Stencil> SIRStencil,
+                                           const std::shared_ptr<SIR> SIR)
     : context_(context), SIRStencil_(SIRStencil), SIR_(SIR) {
   DAWN_LOG(INFO) << "Intializing StencilInstantiation of `" << SIRStencil->Name << "`";
-  DAWN_ASSERT_MSG(SIRStencil, "Stencil does not exist");
```
Comment (twicki) [inline src/dawn/Optimizer/StencilInstantiation.cpp]: Do we still want to check that the shared-ptr does not point to null
`DAWN_ASSERT_MSG(SIRStencil.get(), "Stencil does not exist");`

File: test/unit-test/dawn/Optimizer/Passes/TestFieldAccessIntervals.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    return optimizer->getStencilInstantiationMap()["compute_extent_test_stencil"]->getStencils();
+  }
+};
+
+TEST_F(TestFieldAccessIntervals, test_field_access_interval_01) {
+  auto stencils = loadTest("test_field_access_interval_01.sir");
+  ASSERT_TRUE((stencils.size() == 1));
+  std::shared_ptr<Stencil> stencil = stencils[0];
+
+  ASSERT_TRUE((stencil->getNumStages() == 2));
```
Comment (twicki) [inline test/unit-test/dawn/Optimizer/Passes/TestFieldAccessIntervals.cpp]: why is the extent calculation still here? Don't we do this in TestPAssComputeStageExtents?

File: test/unit-test/dawn/Optimizer/Passes/TestFieldAccessIntervals.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
```
Comment (twicki) [inline test/unit-test/dawn/Optimizer/Passes/TestFieldAccessIntervals.cpp]: as a general note, I had to change this kind of setup slightly for boundary condition checks and am now returning a unique-ptr to the ISIR rather then the vector of stencils. Maybe we should be doing that throughout?

File: src/dawn/Optimizer/StencilInstantiation.cpp (hunk 1 above)
Comment (cosunae) [reply src/dawn/Optimizer/StencilInstantiation.cpp]: done

File: test/unit-test/dawn/Optimizer/Passes/TestFieldAccessIntervals.cpp (hunk 2 above)
Comment (cosunae) [reply test/unit-test/dawn/Optimizer/Passes/TestFieldAccessIntervals.cpp]: yes, but the input data set is different. Just to double check there is nothing weird?

Comment (twicki) [reply test/unit-test/dawn/Optimizer/Passes/TestFieldAccessIntervals.cpp]: We can keep it if we assume that the serialized files might be corrupted.  It is an extra layer of safety but I'm not sure if it belongs into a unittest?
PassComputeStageExtents checks that the extent-calculation is correct and ../../SIR/TestSIRSerializer checks that the de-/ serialization works.

File: test/unit-test/dawn/Optimizer/Passes/TestFieldAccessIntervals.cpp (hunk 3 above)
Comment (cosunae) [reply test/unit-test/dawn/Optimizer/Passes/TestFieldAccessIntervals.cpp]: done

**Comment by cosunae** [comment]
launch jenkins
